### DISCUSSION TITLE
🐛 Bug: Immediate Authorization Failure Using Postgres Docker Compose

### DISCUSSION THREAD

**Discussion by shaydin** [nocodb/nocodb#4907] [Q&A]
### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Steps to reproduce ?

- Using the provided Docker Compose file for Postgres, attempt to spin up a Noco Docker stack.
- Containers will create, start, and show healthy
- Containers and stack look good in Portainer
- Attempt to access <IPADDRESS>:8080 and get met with This Site Can't Be Reached error
- View logs in Portainer for nocodb_nocodb_1 container and see password auth failures for the user/pw in the untouched docker-compose file
- This happens with modified usernames and/or passwords with and without special characters. 

### Desired Behavior

At minimum, the provided docker compose file should spin up successfully and without auth failures. Ideally updating the credentials in the compose file will yield a successful deployment with the ability to begin using the app.

### Project Details

NocoDB used as docker: true
NocoDB Version: version pulled from docker compose
Database used: pg
OS on which NocoDB is running: Debian 11, Docker 20.10.22
All other info N/A
Log:
App started successfully.
Visit -> http://localhost:8080/dashboard
]╔══════════════════════════════════════════════════════════════════════════════╗║                                                                              ║║   EXCEPTION OCCURED!! in PGClient @ createDatabaseIfNotExists                ║║   - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -    ║║   - - - - - - - - - - - - - - - - - - - - - -                                ║║   MESSAGE:                                                                   ║║   password authentication failed for user "postgres"                         ║║                                                                              ║║   - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -    ║║   - - - - - - - - - - - - - - - - - - - - - -                                ║║   CODE:                                                                      ║║   28P01                                                                      ║║                                                                              ║║   - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -    ║║   - - - - - - - - - - - - - - - - - - - - - -                                ║║   STACK:                                                                     ║║   error: password authentication failed for user "postgres"                  ║║   at Parser.parseErrorMessage                                                ║║   (/usr/src/app/node_modules/pg-protocol/dist/parser.js:287:98)              ║║   at Parser.handlePacket                                                     ║║   (/usr/src/app/node_modules/pg-protocol/dist/parser.js:126:29)              ║║   at Parser.parse                                                            ║║   (/usr/src/app/node_modules/pg-protocol/dist/parser.js:39:38)               ║║   at Socket.<anonymous>                                                      ║║   (/usr/src/app/node_modules/pg-protocol/dist/index.js:11:42)                ║║   at Socket.emit (node:events:513:28)                                        ║║   at Socket.emit (node:domain:489:12)                                        ║║   at addChunk (node:internal/streams/readable:315:12)                        ║║   at readableAddChunk (node:internal/streams/readable:289:9)                 ║║   at Socket.Readable.push (node:internal/streams/readable:228:10)            ║║   at TCP.onStreamRead (node:internal/stream_base_commons:190:23)             ║║                                                                              ║║   - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -    ║║   - - - - - - - - - - - - - - - - - - - - - -                                ║║                                                                              ║║                                                                              ║╚══════════════════════════════════════════════════════════════════════════════╝
error: password authentication failed for user "postgres"
    at Parser.parseErrorMessage (/usr/src/app/node_modules/pg-protocol/dist/parser.js:287:98)
    at Parser.handlePacket (/usr/src/app/node_modules/pg-protocol/dist/parser.js:126:29)
    at Parser.parse (/usr/src/app/node_modules/pg-protocol/dist/parser.js:39:38)
    at Socket.<anonymous> (/usr/src/app/node_modules/pg-protocol/dist/index.js:11:42)
    at Socket.emit (node:events:513:28)
    at Socket.emit (node:domain:489:12)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Socket.Readable.push (node:internal/streams/readable:228:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
  length: 104,
  severity: 'FATAL',
  code: '28P01',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'auth.c',
  line: '326',
  routine: 'auth_failed'
}
error: password authentication failed for user "postgres"
    at Parser.parseErrorMessage (/usr/src/app/node_modules/pg-protocol/dist/parser.js:287:98)
    at Parser.handlePacket (/usr/src/app/node_modules/pg-protocol/dist/parser.js:126:29)
    at Parser.parse (/usr/src/app/node_modules/pg-protocol/dist/parser.js:39:38)
    at Socket.<anonymous> (/usr/src/app/node_modules/pg-protocol/dist/index.js:11:42)
    at Socket.emit (node:events:513:28)
    at Socket.emit (node:domain:489:12)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Socket.Readable.push (node:internal/streams/readable:228:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
  length: 104,
  severity: 'FATAL',
  code: '28P01',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'auth.c',
  line: '326',
  routine: 'auth_failed'
}

**Comment by wingkwong** [comment]
I just use that docker-compose.yml to start locally, it is working fine. I couldn't see that error. When you modified username and password (namely `POSTGRES_USER` and `POSTGRES_PASSWORD`), did you also modified `NC_DB` as well?

**Comment by shaydin** [comment]
I initially did not modify NC_DB, and later did modify it to match what I had customized. But again, I see this error on a fresh download of the pg version of the docker-compose file no matter what I do.

**Comment by wingkwong** [comment]
Have you added `-v` when you call `docker-compose down` after changing the password? Besides, you may try to connect to your postgres to check if it is still using the old password. If so, that means you are still reading the old env. Try to add `-v` flag to remove the volume and recreate the container.

**Comment by o1lab** [comment]
Moving this to discussion

**Comment by shaydin** [comment]
Again, this is happening with the unmodified docker compose file, even with no other postgres instance. I'm not adding anything, and I do not have any other postgres DBs running at the moment. I am changing _nothing_ from the supplied file and I'm getting this error. The only postgres env running is the one generated by this docker compose file.

**Comment by haranobuhardo** [comment]
I faced the same issue now. I'm using NocoDB for my internal use case, and use the same default docker-compose.yml username and password for postgre database, but after running it for a couple of months, this issue comes up. I've checked the docker logs (NocoDB container), and it is full of auth error (error: password authentication failed for user "postgres")

hope the dev team can help us to solve this issue. We just started to love NocoDB.

**Reply by wingkwong** [reply]
@haranobuhardo can you provide the following info?

- project info
- did the issue come up after upgrade?
- have you checked if you can use that username and password to log on your postgres database directly?

**Reply by haranobuhardo** [reply]
@wingkwong here are the details:

- project: using docker compose
- Not really sure, but before I upgraded to NocoDB 0.2, the error didn't came up as often as now (but I didn't check the log, I thought it was network issue) 
- yes, I mapped the postgresql db to host machine, and tried to test the connection with DBEaver, and all good (no problem). I also have checked docker logs for the root_db, but there's no fishy error (actually, the username and password of the db are still the same as the default value from docker-compose.yml file in the repo)

**Reply by wingkwong** [reply]
Can you share the exact version number that you are using?

**Reply by haranobuhardo** [reply]
I'm using the latest version of NocoDB (0.202.4)

**Reply by haranobuhardo** [reply]
UPDATE:

I'm not really sure, but I think I just solved this issue.

Here's what I did:
- Renamed the PostgreSQL container's name from _root_db_ to something else like _root_db_db_ in the [docker-compose.yml](https://github.com/nocodb/nocodb/blob/develop/docker-compose/pg/docker-compose.yml)

Idk what's the problem, but recreating the Postgre docker container with the default name didn't solve the issue (also recreating the vol) (but I haven't check if this is caused by broken/corrupted postgresql docker image, but I did updated the psql docker image in the past).

Maybe this is caused by the same psql host and db name? which are "root_db"? idk, this is just my hypothesis.

**Reply by GideonHallett** [reply]
Could you share the changes you made to the docker-compose.yml? - I tried changing the container name in the compose and it made no difference.

We had a NocoDB instance that ran fine for about two months (Ubuntu Server 22.04, current NocoDB, docker.io 24.0.5) and then stopped with exactly this issue; we could connect to the Postgres DB from the CLI with the set credentials just fine, but the nocodb container would not connect.

As an experiment, rolled a brand new machine, everything fresh, tried firing it up, and it gave the error straight off.

As a further experiment, I pruned and removed the Postgres containers and volumes on the new machine, tried installing the MySQL container - and that gave exactly the same issue - the nocodb container started and failed:

       ║╚══════════════════════════════════════════════════════════════════════════════╝
nocodb_1   | Error: Access denied for user 'noco'@'172.21.0.3' (using password: YES)
nocodb_1   |     at Packet.asError (/usr/src/app/node_modules/mysql2/lib/packets/packet.js:728:17)
nocodb_1   |     at ClientHandshake.execute (/usr/src/app/node_modules/mysql2/lib/commands/command.js:29:26)
nocodb_1   |     at Connection.handlePacket (/usr/src/app/node_modules/mysql2/lib/connection.js:478:34)
nocodb_1   |     at PacketParser.onPacket (/usr/src/app/node_modules/mysql2/lib/connection.js:97:12)
nocodb_1   |     at PacketParser.executeStart (/usr/src/app/node_modules/mysql2/lib/packet_parser.js:75:16)
nocodb_1   |     at Socket.<anonymous> (/usr/src/app/node_modules/mysql2/lib/connection.js:104:25)
nocodb_1   |     at Socket.emit (node:events:517:28)
nocodb_1   |     at Socket.emit (node:domain:489:12)
nocodb_1   |     at addChunk (node:internal/streams/readable:335:12)
nocodb_1   |     at readableAddChunk (node:internal/streams/readable:308:9) {
nocodb_1   |   code: 'ER_ACCESS_DENIED_ERROR',
nocodb_1   |   errno: 1045,
nocodb_1   |   sqlState: '28000',
nocodb_1   |   sqlMessage: "Access denied for user 'noco'@'172.21.0.3' (using password: YES)",
nocodb_1   |   sql: undefined

Again, DB access from the command line works fine.

Any tips would be appreciated!

**Reply by haranobuhardo** [reply]
I literally just changed the **root_db** container's name the official docker-compose (with PostgreSQL as the DB) file. But in my case, sometimes the NocoDB container would still connect to the DB, but after sometimes it wouldn't connect for a while. And this cycle keep going on and on.

But looks like in your case, the NocoDB container can't connect with the PostgreDB container all the time? Have you tried to ping the DB from NocoDB container?

**Reply by GideonHallett** [reply]
No, I haven't - the NocoDB container just restarts itself endlessly; as part of the startup it tries to connect to the root_db_1 container, fails, resets and tries again; uptime is never long enough to get a shell on the container.

One thing that is possibly relevant; the issue originally manifested after the Ubuntu server was updated from 22.04.2 to 22.04.3; I'll try installing on top of 22.04.2 and see what happens.

**Reply by GideonHallett** [reply]
Well, scratch that hypothesis. Built a 22.04.2 server, installed the same software, cloned nocodb and:

PG container (with changed container name!):

root_newdb_1  | 2024-01-04 10:23:24.962 UTC [47] LOG:  database system is shut down
root_newdb_1  |  done
root_newdb_1  | server stopped
root_newdb_1  | 
root_newdb_1  | PostgreSQL init process complete; ready for start up.
root_newdb_1  | 
root_newdb_1  | 2024-01-04 10:23:25.045 UTC [1] LOG:  starting PostgreSQL 16.1 (Debian 16.1-1.pgdg120+1) on x86_64-pc-linux-gnu,
 compiled by gcc (Debian 12.2.0-14) 12.2.0, 64-bit
root_newdb_1  | 2024-01-04 10:23:25.045 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
root_newdb_1  | 2024-01-04 10:23:25.045 UTC [1] LOG:  listening on IPv6 address "::", port 5432
root_newdb_1  | 2024-01-04 10:23:25.048 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
root_newdb_1  | 2024-01-04 10:23:25.052 UTC [63] LOG:  database system was shut down at 2024-01-04 10:23:24 UTC
root_newdb_1  | 2024-01-04 10:23:25.057 UTC [1] LOG:  database system is ready to accept connections
root_newdb_1  | 2024-01-04 10:23:39.775 UTC [75] FATAL:  password authentication failed for user "postgres"
root_newdb_1  | 2024-01-04 10:23:39.775 UTC [75] DETAIL:  Connection matched file "/var/lib/postgresql/data/pg_hba.conf" line 12
8: "host all all all scram-sha-256"
root_newdb_1  | 2024-01-04 10:23:39.839 UTC [76] FATAL:  password authentication failed for user "postgres"
root_newdb_1  | 2024-01-04 10:23:39.839 UTC [76] DETAIL:  Connection matched file "/var/lib/postgresql/data/pg_hba.conf" line 12
8: "host all all all scram-sha-256"
root_newdb_1  | 2024-01-04 10:23:44.223 UTC [84] FATAL:  password authentication failed for user "postgres"

nocodb container:

nocodb_1      | (node:12) NOTE: We are formalizing our plans to enter AWS SDK for JavaScript (v2) into maintenance mode in 2023.
nocodb_1      | 
nocodb_1      | Please migrate your code to use AWS SDK for JavaScript (v3).
nocodb_1      | For more information, check the migration guide at https://a.co/7PzMCcy
nocodb_1      | (Use `node --trace-warnings ...` to show where the warning was created)
nocodb_1      | ╔══════════════════════════════════════════════════════════════════════════════╗║                               
                                               ║║   EXCEPTION OCCURED!! in PGClient @ createDatabaseIfNotExists                ║║   - -
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -    ║║   - - - - - - - - - - - - - - - - - - - - - -               
                 ║║   MESSAGE:                                                                   ║║   password authentication failed fo
r user "postgres"                         ║║                                                                              ║║   - - - - 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -    ║║   - - - - - - - - - - - - - - - - - - - - - -                    
            ║║   CODE:                                                                      ║║   28P01                                 
                                     ║║                                                                              ║║   - - - - - - -
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - -    ║║   - - - - - - - - - - - - - - - - - - - - - -                         
       ║║   STACK:                                                                     ║║   error: password authentication failed for u
ser "postgres"                  ║║   at Parser.parseErrorMessage                                                ║║   (/usr/src/app/node
_modules/pg-protocol/dist/parser.js:287:98)              ║║   at Parser.handlePacket                                                   
  ║║   (/usr/src/app/node_modules/pg-protocol/dist/parser.js:126:29)              ║║   at Parser.parse                                 
                           ║║   (/usr/src/app/node_modules/pg-protocol/dist/parser.js:39:38)               ║║   at Socket.<anonymous>  
                                                    ║║   (/usr/src/app/node_modules/pg-protocol/dist/index.js:11:42)                ║║ 
  at Socket.emit (node:events:517:28)                                        ║║   at Socket.emit (node:domain:489:12)                  
                      ║║   at addChunk (node:internal/streams/readable:335:12)                        ║║   at readableAddChunk (node:in
ternal/streams/readable:308:9)                 ║║   at Readable.push (node:internal/streams/readable:245:10)                   ║║   at 
TCP.onStreamRead (node:internal/stream_base_commons:190:23)             ║║                                                             
                 ║║   - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -    ║║   - - - - - - - - - - - - - - - - -
 - - - - -                                ║║                                                                              ║║           
                                                                   ║╚══════════════════════════════════════════════════════════════════
════════════╝
nocodb_1      | error: password authentication failed for user "postgres"
nocodb_1      |     at Parser.parseErrorMessage (/usr/src/app/node_modules/pg-protocol/dist/parser.js:287:98)
nocodb_1      |     at Parser.handlePacket (/usr/src/app/node_modules/pg-protocol/dist/parser.js:126:29)
nocodb_1      |     at Parser.parse (/usr/src/app/node_modules/pg-protocol/dist/parser.js:39:38)
nocodb_1      |     at Socket.<anonymous> (/usr/src/app/node_modules/pg-protocol/dist/index.js:11:42)
nocodb_1      |     at Socket.emit (node:events:517:28)
nocodb_1      |     at Socket.emit (node:domain:489:12)
nocodb_1      |     at addChunk (node:internal/streams/readable:335:12)
nocodb_1      |     at readableAddChunk (node:internal/streams/readable:308:9)
nocodb_1      |     at Readable.push (node:internal/streams/readable:245:10)
nocodb_1      |     at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
nocodb_1      |   length: 104,
nocodb_1      |   severity: 'FATAL',
nocodb_1      |   code: '28P01',
nocodb_1      |   detail: undefined,
nocodb_1      |   hint: undefined,
nocodb_1      |   position: undefined,
nocodb_1      |   internalPosition: undefined,
nocodb_1      |   internalQuery: undefined,
nocodb_1      |   where: undefined,
nocodb_1      |   schema: undefined,
nocodb_1      |   table: undefined,
nocodb_1      |   column: undefined,
nocodb_1      |   dataType: undefined,
nocodb_1      |   constraint: undefined,
nocodb_1      |   file: 'auth.c',
nocodb_1      |   line: '323',
nocodb_1      |   routine: 'auth_failed'
nocodb_1      | }
nocodb_1      | [Nest] 12  - 01/04/2024, 10:23:39 AM     LOG [NestFactory] Starting Nest applica
tion...
nocodb_1      | [Nest] 12  - 01/04/2024, 10:23:39 AM     LOG [InstanceLoader] EventEmitterModule
 dependencies initialized +1ms
nocodb_1      | [Nest] 12  - 01/04/2024, 10:23:39 AM     LOG [InstanceLoader] PassportModule dep
endencies initialized +0ms
nocodb_1      | [Nest] 12  - 01/04/2024, 10:23:39 AM     LOG [InstanceLoader] MulterModule depen
dencies initialized +0ms
nocodb_1      | [Nest] 12  - 01/04/2024, 10:23:39 AM     LOG [InstanceLoader] MulterModule depen
dencies initialized +0ms
nocodb_1      | [Nest] 12  - 01/04/2024, 10:23:39 AM     LOG [InstanceLoader] ConfigHostModule d
ependencies initialized +1ms
nocodb_1      | [Nest] 12  - 01/04/2024, 10:23:39 AM     LOG [InstanceLoader] DiscoveryModule de
pendencies initialized +0ms
nocodb_1      | [Nest] 12  - 01/04/2024, 10:23:39 AM     LOG [InstanceLoader] ConfigModule depen
dencies initialized +0ms
nocodb_1      | [Nest] 12  - 01/04/2024, 10:23:39 AM     LOG [InstanceLoader] EventEmitterModule
 dependencies initialized +0ms
nocodb_1      | [Nest] 12  - 01/04/2024, 10:23:39 AM   ERROR [ExceptionHandler] error: password 
authentication failed for user "postgres"
nocodb_1      | Error: error: password authentication failed for user "postgres"
nocodb_1      |     at NcConfig.create (/usr/src/app/docker/main.js:2:1892717)
nocodb_1      |     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
nocodb_1      |     at async InstanceWrapper.useFactory [as metatype] (/usr/src/app/docker/main.js:2:1360883)
nocodb_1      |     at async Injector.instantiateClass (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:369:37)
nocodb_1      |     at async callback (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:65:34)
nocodb_1      |     at async Injector.resolveConstructorParams (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:144:
24)
nocodb_1      |     at async Injector.loadInstance (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:70:13)
nocodb_1      |     at async Injector.loadProvider (/usr/src/app/node_modules/@nestjs/core/injector/injector.js:97:9)
nocodb_1      |     at async /usr/src/app/node_modules/@nestjs/core/injector/instance-loader.js:56:13
nocodb_1      |     at async Promise.all (index 3)

Annoying - this is a fresh machine, with no legacy cruft in it, but nocodb just doesn't want to play.

**Reply by haranobuhardo** [reply]
@GideonHallett 

`root_newdb_1 | 2024-01-04 10:23:39.839 UTC [76] FATAL: password authentication failed for user "postgres"`

Can you make sure your postgresql are correctly setup? Because in my case, my postgre container doesn't show any error (like above)

And for nocodb container, I don't think it will terminated if you failed it failed to connect with the db, it will just keep retrying to connect to the db.

**Reply by GideonHallett** [reply]
@haranobuhardo I'm reasonably sure that the Postgresql container is more-or-less OK - it's up and stable; I can connect to it via a Postgres client on the command line; I tried changing the pg_hba.conf in the container to us MD5 rather than SCRAM and it made no difference.

It's only when the nocodb container tries to connect that the DB container gives an error.

For the nocodb container, this is pretty typical for docker ps:

CONTAINER ID   IMAGE                  COMMAND                  CREATED        STATUS                         PORTS      NAMES
14c0d5d5af35   nocodb/nocodb:latest   "/usr/bin/dumb-init …"   22 hours ago   Restarting (1) 2 seconds ago              pg_nocodb_1
7647cf51ecba   postgres               "docker-entrypoint.s…"   22 hours ago   Up 22 hours (healthy)          5432/tcp   pg_root_newdb_1

Does the same with MySQL, too:

CONTAINER ID   IMAGE                  COMMAND                  CREATED         STATUS                   PORTS                                       NAMES
fd68a7426088   nocodb/nocodb:latest   "/usr/bin/dumb-init …"   7 minutes ago   Up 2 seconds             0.0.0.0:8080->8080/tcp, :::8080->8080/tcp   mysql_nocodb_1
2f7e2a783f10   mysql:8.0.35           "docker-entrypoint.s…"   7 minutes ago   Up 7 minutes (healthy)   3306/tcp, 33060/tcp                         mysql_root_db_1

(Again, CLI access to the DB container works fine.)

The common factors in both cases are Docker and the nocodb container.

(FWIW, 'curl http://get.nocodb.com/linux-x64 -o nocodb -L && chmod +x nocodb  && ./nocodb' seems to work OK - so it might be easier to go with node.js rather than Docker!)

**Reply by Deedu** [reply]
**Progress, not fix...**
It didn't make the UI accessible for me, but manually changing the password of the Postgres user in the database container stopped the nocodb container from being on an infinite restart loop.

To change the default password manually:

1.  Open docker desktop
2. Go to container called nocodb-db-1 _(may appear as db-1 in preview)_
3. Click the exec tab (gives you a shell in the container)
4. enter command `psql -U postgres`
5. then enter command `\password postgres`
6.  --> you will be prompted to enter a password. You can get this password from your env file, nocodb/docker.env, copy the db password from there

I still haven't been able to get nocodb running. Hopefullly this helps someone else.. @shaydin @GideonHallett

**Reply by effetmonstre** [reply]
I was able to bypass this error by installing everything as root
